File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int pre = 1, post = 1;

        vector<int> res;

        for(int i: nums) {
            res.push_back(pre);
            pre *= i;
        }

        for(int i = nums.size() - 1; i >= 0; i--) {
            res[i] *= post;
            post = post*nums[i];
        }

        return res;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pre = [1]*(len(nums)+1)
        post = [1]*(len(nums)+1)

        for i in range(0, len(nums)):
            pre[i+1] = pre[i]*nums[i]
        
        for i in range(len(nums) - 1, -1, -1):
            post[i-1] = post[i]*nums[i]

        for i in range(len(nums)):
            nums[i] = pre[i]*post[i]

        return nums

